### DISCUSSION TITLE
After deploying Subgen, I encountered a problem with connecting it to Emby.

### DISCUSSION THREAD

**Discussion by milesxia** [McCloudS/subgen#115] [Ideas]
I deployed a container on my QNAP NAS using Docker Compose. I usually use Emby as my media player. However, I've encountered an issue: after entering the local IP address and port number of Subgen (http://192.168.200.132:9000/emby) in the Webhooks settings of Emby and clicking the test link, it says the URL is incorrect. But I can access it through a web browser. Is there any solution to this problem?

**Comment by McCloudS** [comment]
I don’t have a valid Emby subscription with webhooks at this time to see what the test button actually does.  Did you actually try playing or adding a video after saving the webhook in Emby and checking the subgen logs?

it could also depend on your docker configuration, if your containers don’t route to each other.  Can you ping subgen from inside the Emby container?

**Reply by milesxia** [reply]
I have tried playing movies without subtitles, but Subgen did not start working, and there was no additional content in the logs. I tried several movies (both in English and Japanese). I apologize for my lack of technical skills; I don't know how to ping Subgen within Emby. As far as I understand, clicking the "Test Connection" button is supposed to test whether Subgen can connect, but they are not working together.

**Reply by milesxia** [reply]
INFO:     172.29.36.1:49886 - "HEAD /emby HTTP/1.1" 405 Method Not Allowed
INFO:     172.29.36.1:56572 - "GET /emby HTTP/1.1" 200 OK
INFO:     172.29.36.1:57460 - "POST /emby HTTP/1.1" 200 OK
WARNING:  Invalid HTTP request received.

**Reply by McCloudS** [reply]
Emby may have changed the structure of the message when they moved from webhooks to 'notifications'.  I'll see what I can check out.

**Reply by milesxia** [reply]
Thank you for making the modifications to the container for me. After redeploying it with the updated image, the logs no longer show errors, which is great. However, in Emby, I still see the message "An error occurred while sending the Webhook test message. Please check the URL you entered and try again." I also tried creating a Jellyfin plugin by following the documentation, but in the end, I still don't know how to use it. I'm really sorry, but at this point, I'm not sure how to describe the issues I'm facing. Once again, I sincerely thank you for all the work you've done for me.

**Comment by McCloudS** [comment]
I made some changes to not throw an error on the emby test connection, but everything else appears to be working fine as is.  If you re-pull the docker image or just UPDATE=True in your docker env variables.  You also need to make sure that subgen has access to the same media directories as Emby at the same location or use path mapping settings in subgen.

`2024-09-05 08:00:14,172 DEBUG: Raw response: {"Title":"Scott is playing Billy the Kid - S2, Ep1 - The Road to Hell on Chrome macOS","Date":"2024-09-05T14:00:14.1285550Z","Event":"playback.start","User":{"Name":"Scott","Id":"961f390f0306415687d24263559594ab"},"Item":{"Name":"The Road to Hell","ServerId":"2965b319f8a849ebb156e4ea01e9a5c5","Id":"478","DateCreated":"2023-10-16T04:00:00.0000000Z","Container":"mkv","SortName":"Road to Hell","PremiereDate":"2023-10-15T07:00:00.0000000Z","ExternalUrls":[{"Name":"TheTVDB","Url":"https://thetvdb.com/?tab=episode&id=9944695"}],"Path":"/mnt/user/TV/Billy the Kid/Season 02/Billy the Kid - S02E01 - The Road to Hell.mkv","Overview":"The Lincoln County War is on the verge of kicking off in a big way. Billy has joined Englishman John Tunstall in his commercial battle with Major Murphy and the House, but even bigger players lurk in the background and the stakes are high.","Taglines":[],"Genres":[],"RunTimeTicks":28512110000,"Size":275927118,"FileName":"Billy the Kid - S02E01 - The Road to Hell.mkv","Bitrate":774203,"ProductionYear":2023,"IndexNumber":1,"ParentIndexNumber":2,"RemoteTrailers":[],"ProviderIds":{"Tvdb":"9944695"},"IsFolder":false,"ParentId":"477","Type":"Episode","Studios":[],"GenreItems":[],"TagItems":[],"ParentLogoItemId":"44","ParentBackdropItemId":"44","ParentBackdropImageTags":["ee06fe740a11ec9944c619e9f95d86ea"],"SeriesName":"Billy the Kid","SeriesId":"44","SeasonId":"477","PrimaryImageAspectRatio":1.7777777777777777,"SeriesPrimaryImageTag":"ebc8cb7fe2cd3a97fdfa728a86872bff","SeasonName":"Season 2","ImageTags":{"Primary":"fd5c872d31d9973d331bc2f0dc355a72"},"BackdropImageTags":[],"ParentLogoImageTag":"80d339a1912f731bce717291b4933c0e","ParentThumbItemId":"44","ParentThumbImageTag":"789f8891b1ad000a8cdede46e75e0155","MediaType":"Video","Width":1280,"Height":720},"Server":{"Name":"Tower","Id":"2965b319f8a849ebb156e4ea01e9a5c5","Version":"4.8.7.0"},"Session":{"RemoteEndPoint":"192.168.1.21","Client":"Emby Web","DeviceName":"Chrome macOS","DeviceId":"TW96aWxsYS81LjAgKE1hY2ludG9zaDsgSW50ZWwgTWFjIE9TIFggMTBfMTVfNykgQXBwbGVXZWJLaXQvNTM3LjM2IChLSFRNTCwgbGlrZSBHZWNrbykgQ2hyb21lLzExNy4wLjAuMCBTYWZhcmkvNTM3LjM2fDE2OTgwOTg4NTIwMTI1","ApplicationVersion":"4.8.7.0","Id":"81c1dc32bdec8bcce23056e6668cf267"},"PlaybackInfo":{"PositionTicks":15575287270,"PlaylistIndex":8,"PlaylistLength":16,"PlaySessionId":"b5fbcaa1f76a456aa95f7ffd09ba9299"}}
2024-09-05 08:00:14,172 DEBUG: Emby event detected is: playback.start
2024-09-05 08:00:14,172 DEBUG: Path of file: /mnt/user/TV/Billy the Kid/Season 02/Billy the Kid - S02E01 - The Road to Hell.mkv
2024-09-05 08:00:14,172 DEBUG: Updated path: /Volumes/TV/Billy the Kid/Season 02/Billy the Kid - S02E01 - The Road to Hell.mkv
2024-09-05 08:00:14,354 DEBUG: No subtitles in 'ena' language found in the video.
2024-09-05 08:00:14,419 INFO: Added Billy the Kid - S02E01 - The Road to Hell.mkv for transcription.
2024-09-05 08:00:14,419 INFO: Transcribing file: Billy the Kid - S02E01 - The Road to Hell.mkv
2024-09-05 08:00:14,419 DEBUG: Model was purged, need to re-create
INFO:     192.168.1.111:53780 - "POST /emby HTTP/1.1" 200 OK
2024-09-05 08:01:53,069 INFO: Processing audio with duration 47:31.104
2024-09-05 08:01:55,696 INFO: Detected language 'en' with probability 0.98
Detected Language: english
Transcribe:   5%|██████▊                                                                                                                                        | 136.88/2851.1 [00:09<02:29, 18.16sec/s]`

**Comment by McCloudS** [comment]
My settings look like: 
![image](https://github.com/user-attachments/assets/6712b918-6fca-4072-a7d8-16d3745897e1)
![image](https://github.com/user-attachments/assets/20016131-8872-46f2-a310-3de73af9924e)
![image](https://github.com/user-attachments/assets/39e64de2-2eb7-4186-93b1-f24c66a0691e)

If your "request content type" isn't multipart/form-data, it won't work.

Clicking "Send Test Notification" should show no errors and just say "Notification Sent" in Emby and the below in subgen: `2024-09-06 07:13:55,587 DEBUG: Raw response: {"Title":"Test Notification","Description":"Test Notification Description","Date":"2024-09-06T13:13:55.5965144Z","Event":"system.notificationtest","Server":{"Name":"Tower","Id":"2965b319f8a849ebb156e4ea01e9a5c5","Version":"4.8.8.0"}}
2024-09-06 07:13:55,587 DEBUG: Emby event detected is: system.notificationtest
2024-09-06 07:13:55,587 INFO: Emby test message received!
`

Your troubleshooting skills may be limiting here, since you don't know how to ensure network connectivity between the containers.  The errors kind of point to the containers being unable to see eachother or route traffic to eachother.  Verify you have them in the same QNAP/Docker network.

**Reply by milesxia** [reply]
Thank you for your reply. From your screenshots, I found the issue with my Emby container. My Webhooks configuration lacks the "Request Content Type" option, which I believe is the reason why my Emby Webhooks are unable to send information, leading to the integration issue. Currently, I don't have a solution for this and will have to wait for a future deployment. Once again, thank you for your time and effort.

**Reply by milesxia** [reply]
![Uploading 截屏2024-09-10 10.48.19.png…]()
Hello, I updated my Emby today, and now the relevant options are available in the Webhooks settings. However, I am still facing the issue of a server connection timeout, which is preventing the two containers from interacting. There are no new entries in the Subgen logs either. Both containers are deployed on the same device, within the same local network, and they only differ in port numbers.

**Comment by milesxia** [comment]
![截屏2024-09-10 10 48 19](https://github.com/user-attachments/assets/3fcac31f-ae18-4efc-9a05-2603bf1a505d)

**Reply by milesxia** [reply]
#docker-compose.yml
version: '2'
services:
  subgen:
    container_name: subgen
    tty: true
    image: mccloud/subgen
    environment:
       - "WHISPER_MODEL=medium" # 实测cpu使用medium即可，如果性能不够可以试试small
       - "WHISPER_THREADS=4" # 并发线程数
       - "PROCADDEDMEDIA=True" # 是否在jellyfin新增媒体时触发
       - "PROCMEDIAONPLAY=true" # 是否在jellyfin媒体播放时触发
       - "NAMESUBLANG=chi" # 需要生成的字幕文件语言后缀
       - "SKIPIFINTERNALSUBLANG=chi" # 是否跳过已有中文字幕的视频
       - "SKIPIFEXTERNALSUB=true" # 是否跳过带内嵌字幕的视频
       - "PLEXTOKEN=plextoken"
       - "PLEXSERVER=http://plexserver:32400"
       - "JELLYFINTOKEN="
       - "JELLYFINSERVER="
       - "WEBHOOKPORT=9000" # 服务端口
       - "CONCURRENT_TRANSCRIPTIONS=2" # 同时转录的视频数
       - "WORD_LEVEL_HIGHLIGHT=False"
       - "DEBUG=True" # 开启DEBUG日志，实测无效要改源码
       - "USE_PATH_MAPPING=true" # 开启重映射
       - "PATH_MAPPING_FROM=/media"
       - "PATH_MAPPING_TO=/share/CACHEDEV1_DATA/Container/media/"
       - "FORCE_DETECTED_LANGUAGE_TO=zh" # 强制指定转录语言
       - "TRANSCRIBE_OR_TRANSLATE=transcribe" # transcribe可以实现英文转中文
       - "TRANSCRIBE_DEVICE=cpu" # 使用硬件，有独显可以换成gpu
       - "CLEAR_VRAM_ON_COMPLETE=True" # 开启垃圾收集
       - "MODEL_PATH=/subgen/models" # 模型保存路径，对应下面的映射
       - "UPDATE=true" # 是否自动从git库更新程序，一些最新更改可能没在docker镜像中
       - "APPEND=False" # 在字幕结尾添加whisper文案
       - "USE_MODEL_PROMPT=true" # 在字幕结尾添加whisper文案
       - "CUSTOM_MODEL_PROMPT=你好，欢迎来到我的讲座。" # 自定义提示词
       - "LRC_FOR_AUDIO_FILES=True" # 支持歌曲LRC歌词文件生成
       - "CUSTOM_REGROUP=cm_sl=84_sl=42++++++1" # 优化字幕格式的
       - "SUBGEN_KWARGS={'vad': true, 'prefix': '对，没错，继续。'}" # 通过vad避免幻听和提示前缀避免变回英文
    volumes:
       - "/share/CACHEDEV1_DATA/Container/media/video/link/movie:/movie"
       - "/share/CACHEDEV1_DATA/Container/media/video/link/cartoon:/cartoon"
       - "/share/CACHEDEV1_DATA/Container/media/video/link/tv:/tv"
       - "/share/CACHEDEV1_DATA/Container/subgen/models:/subgen/models"
    ports:
       - "9000:9000"


I deployed using the docker-compose command, and since I'm using Nastool to manage my automation, all my media files are in a hard link state. I wonder if this would have any impact.

**Comment by McCloudS** [comment]
You should be getting something in subgen logs if it was getting hit with requests.  Points to the containers not being able to see each other.

you might be able to try http://subgen:9000/emby and see if docker can route that.  You could also try using the internal docker ip instead of what appears to be your machines LAN address.

does the subgen container show started and you said you can access it from your desktop?

**Reply by milesxia** [reply]
I’ve solved the issue, and now the Webhooks messages are successfully being sent. Thank you for your help during this time. I resolved the issue by redeploying the Docker container and configuring all the environment variables on the webpage at port 9000, which worked!
